### PR TITLE
refactor(components): enforce public imports in ci

### PR DESCRIPTION
**Pull request by akoumpa**
# What does this PR do ?

Enforces imports of configured Automodel components through package-level symbols declared in `__all__`, including recipes and every other runtime consumer under `nemo_automodel`.

# Changelog

- Add a custom Import Linter component-interface contract that scans all runtime modules under `nemo_automodel`.
- Reject private component-submodule imports and component-root names absent from `__all__`.
- Migrate existing recipe, framework, model, CLI, and cross-component imports to their public component interfaces.
- Add lazy package exports to preserve optional-dependency and import-cycle behavior.
- Add focused contract tests and update repository policy documentation.

# Validation

- `uvx --from import-linter==2.4.0 lint-imports --debug --verbose --no-cache` — 1 contract kept, 0 broken.
- `uvx --from "ruff~=0.12.0" ruff check .` — passed.
- `uvx --from "ruff~=0.12.0" ruff format --check .` — 799 files already formatted.
- Focused and upstream-overlap pytest selection — 450 passed.
- `uvx --from bandit bandit -r app.py nemo_automodel examples scripts tools tutorials -t B614 -q` — passed.
- `git diff HEAD^ --check` — passed.

# Before your PR is "Ready for review"

**Pre checks**:

- [x] Make sure you read and followed [Contributor guidelines](https://github.com/NVIDIA-NeMo/Automodel/blob/main/CONTRIBUTING.md)
- [x] Did you write any new necessary tests?
- [x] Did you add or update any necessary documentation?

# Additional Information

- Draft while the full CI matrix runs.

### CODE REVIEW

File: docs/about/key-features.mdx
Comment (jgerh) [inline docs/about/key-features.mdx]:
```suggestion
- **Hugging Face native**: Train any model from the Hub with no checkpoint conversion. Day-0 support for new releases.
```

Comment (jgerh) [inline docs/about/key-features.mdx]:
```suggestion
- **Any scale**: The same recipe runs on 1 GPU or across hundreds of nodes. Parallelism is configuration, not code.
```

Comment (jgerh) [inline docs/about/key-features.mdx]:
```suggestion
Copy the checked-in `slurm.sub` or download it from GitHub (available in the [repository root `slurm.sub` file](https://github.com/NVIDIA-NeMo/Automodel/blob/main/slurm.sub)), edit its `CONFIG` and `#SBATCH` directives for your cluster, and submit it with `sbatch`. See the [Cluster guide](/job-launchers/slurm-cluster).
```

Comment (jgerh) [inline docs/about/key-features.mdx]:
```suggestion
| `moe/` | Mixture-of-Experts routing and DeepEP integration |
```

File: docs/about/key-features.mdx [hunk 1]
```diff
@@ -131,7 +131,7 @@ Components are modular, self-contained building blocks that recipes assemble:
 | `loss/` | Cross-entropy, linear cross-entropy, KD loss |
 | `launcher/` | Interactive, SkyPilot, and NeMo-Run job launch; Slurm uses the root-level `slurm.sub` script |
 
-Each component can be used independently and has no cross-module imports.
+Every component consumer—including recipes and other components—imports symbols from the target component's package API. The package's `__all__` defines that public interface.
```
Comment (jgerh) [inline docs/about/key-features.mdx]:
```suggestion
Every component consumer, including recipes and other components, imports symbols from the target component's package API. The package's `__all__` defines that public interface.
```

File: docs/repository-structure.mdx
Comment (jgerh) [inline docs/repository-structure.mdx]:
```suggestion
description: "Overview of the NeMo AutoModel repository layout, including components, recipes, and the CLI."
```

Comment (jgerh) [inline docs/repository-structure.mdx]:
```suggestion
The NeMo AutoModel source code is available under the [`nemo_automodel`](https://github.com/NVIDIA-NeMo/Automodel/tree/main/nemo_automodel) directory. It is organized into three directories:
- [`components/`](https://github.com/NVIDIA-NeMo/Automodel/tree/main/nemo_automodel/components) - Self-contained modules
```

Comment (jgerh) [inline docs/repository-structure.mdx]:
```suggestion
- [`cli/`](https://github.com/NVIDIA-NeMo/Automodel/tree/main/nemo_automodel/cli) - CLI entry-point and job launcher dispatch
```

Comment (jgerh) [inline docs/repository-structure.mdx]:
```suggestion
The command fine-tunes the Llama3.2-1B model on the SQuAD dataset with two GPUs using the [`llama3_2_1b_squad.yaml`](https://github.com/NVIDIA-NeMo/Automodel/blob/main/examples/llm_finetune/llama3_2/llama3_2_1b_squad.yaml) config.
```

Comment (jgerh) [inline docs/repository-structure.mdx]:
```suggestion
An example YAML configuration is shown below. The complete config is available in the [Llama 3.2 SQuAD example config](https://github.com/NVIDIA-NeMo/Automodel/blob/main/examples/llm_finetune/llama3_2/llama3_2_1b_squad.yaml):
```

**Review by jgerh — COMMENTED** [review]
Completed tech pubs review of markdown files and provided a few copyedits and suggested text revisions.

File: nemo_automodel/recipes/llm/train_ft.py [hunk 2]
```diff
@@ -375,7 +388,7 @@ def _build_pp_collate_wrapper(cfg_model, pp_enabled: bool):
         )
         return None
 
-    from nemo_automodel.components.datasets.utils import add_causal_masks_to_batch
+    from nemo_automodel.components.datasets import add_causal_masks_to_batch
```
Comment (yuhezhang-ai) [inline nemo_automodel/recipes/llm/train_ft.py:391]: **[P1] Update the remaining patches to this package-level target.** tests/unit_tests/recipes/test_train_ft.py:1909,1935,1958 still patches nemo_automodel.components.datasets.utils.add_causal_masks_to_batch. Because datasets.__getattr__ caches the resolved object in globals(), that submodule mock can either be cached in the package namespace beyond teardown or be ignored if the package attribute was materialized earlier, making the tests order-dependent. Please patch nemo_automodel.components.datasets.add_causal_masks_to_batch instead. The same stale-target pattern remains in test_diffusion_train_metrics.py:299 and _diffusers/test_auto_diffusion_pipeline.py:1204.

File: tests/unit_tests/test_component_imports.py [hunk 3]
```diff
@@ -0,0 +1,173 @@
...
+# distributed under the License is distributed on an "AS IS" BASIS,
+# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+# See the License for the specific language governing permissions and
+# limitations under the License.
+
+from pathlib import Path
+from unittest.mock import Mock
+
+from tools.component_imports import ComponentImport, find_component_imports
+from tools.import_linter_contracts import ComponentInterfaceContract
```
Comment (yuhezhang-ai) [inline tests/unit_tests/test_component_imports.py:19]: **[P1] Keep default unit-test collection independent of lint-only dependencies.** This unconditional import pulls in grimp and importlinter, but those packages are only in the linting dependency group while the default groups are build, docs, and test. A documented uv sync --frozen followed by pytest tests/unit_tests/ therefore fails during collection. Please either move the dependencies into the test group or isolate/conditionally import only the contract-specific tests; a module-level skip would also hide the scanner tests that do not need these packages.

File: tools/import_linter_contracts.py [hunk 4]
```diff
@@ -0,0 +1,92 @@
...
+    def check(self, graph: ImportGraph, verbose: bool) -> ContractCheck:
+        """Allow direct component edges only when they use package exports."""
+        modules = {
+            module.name
+            for module in module_expressions_to_modules(graph, self.modules)  # type: ignore[arg-type]
+        }
+        component_imports = find_component_imports(_PROJECT_ROOT, modules)
+
+        for component_import in component_imports:
+            self._remove_direct_import(graph, component_import)
```
Comment (yuhezhang-ai) [inline tools/import_linter_contracts.py:48]: **[P1] Preserve the independence check on the original graph.** This removes every scanner-discovered edge into a configured component before super().check() runs. Any chain between configured components must end on one of those edges, so the inherited invalid_chains check cannot detect it; consumer-to-component ignore_imports entries also become unmatched. For example, a loss -> moe -> distributed dependency that main rejects is kept here. Please run the independence check on the unmodified graph and combine it with the interface scan. If interface-only enforcement is intentional, derive directly from Contract and remove the dead chain/ignore machinery instead.

File: nemo_automodel/components/distributed/__init__.py [hunk 5]
```diff
@@ -35,5 +42,106 @@
...
+    ),
+    "translate_to_lora": (".parallel_styles", "translate_to_lora"),
+    "unshard_context_parallel_tensor": (".context_parallel.utils", "unshard_context_parallel_tensor"),
+    "unwrap_checkpoint_wrapper": (".activation_checkpointing", "unwrap_checkpoint_wrapper"),
+}
+
+__all__ += sorted(_LAZY_ATTRS.keys())
+
+
+def __getattr__(name: str) -> object:
```
Comment (yuhezhang-ai) [inline nemo_automodel/components/distributed/__init__.py:134]: **[P1] Preserve static types for lazy exports.** With __getattr__(name: str) -> object, imports of every name known only through _LAZY_ATTRS are typed as object; this pattern affects roughly 250 exports across the component packages. The strict directories stay green only because selected names are declared under TYPE_CHECKING or imported eagerly. Please provide TYPE_CHECKING declarations or stubs for every lazy export and add a parity check between those declarations and _LAZY_ATTRS.keys().

File: tools/component_imports.py [hunk 6]
```diff
@@ -0,0 +1,345 @@
...
+            isinstance(statement, ast.AugAssign)
+            and isinstance(statement.target, ast.Name)
+            and statement.target.id == "__all__"
+            and isinstance(statement.op, ast.Add)
+        ):
+            value = _static_value(statement.value, values)
+            if value is not _UNKNOWN:
+                exports.update(_string_set(value))
+
+    for node in ast.walk(tree):
```
Comment (yuhezhang-ai) [inline tools/component_imports.py:264]: **[P1] Evaluate __all__ only at module level and in execution order.** Walking the whole tree collects __all__.append/extend calls inside functions, unreachable branches, and statements that occur before a later reset, so the scanner can approve a symbol the package does not export at runtime. There is a related state bug above: __all__ += ... updates exports but not values["__all__"], so a later __all__ = sorted(__all__) drops the additions. Please process supported top-level statements in order, keep the tracked value synchronized, reject unsupported shapes with a targeted path/line error, and cover nested/conditional/reset/AugAssign cases.

File: docs/about/key-features.mdx (hunk 1 above)
Comment (yuhezhang-ai) [inline docs/about/key-features.mdx]: **[P2] Scope this claim to what CI actually enforces.** The contract config covers 11 of the 19 top-level component packages; _peft, attention, cuda_graphs, eval, models, moe, quantization, and speculative are omitted, and migrated code still imports private modules from some of them. Please either extend the configured scope or narrow this wording to the component packages currently covered.

File: nemo_automodel/components/checkpoint/__init__.py [hunk 7]
```diff
@@ -22,3 +24,37 @@
 
 if _is_geq_torch_2_9():
     _nemo__apply_async_patch()
+
+_LAZY_ATTRS = {
+    "Checkpointer": (".checkpointing", "Checkpointer"),
+    "StateDictAdapter": (".state_dict_adapter", "StateDictAdapter"),
+    "find_latest_checkpoint": (".utils", "find_latest_checkpoint"),
+    "get_checkpoint_tensor_dtypes": (".utils", "_get_checkpoint_tensor_dtypes"),
```
Comment (yuhezhang-ai) [inline nemo_automodel/components/checkpoint/__init__.py:32]: **[P2] Avoid publishing private helpers by aliasing them.** This exposes a public name backed by a private implementation, and the same pattern is repeated for a number of helpers that consumers then alias back to underscore-prefixed names. Since adding a name to __all__ creates a long-lived API obligation, please either promote and rename the implementation consistently or route callers through an existing public operation instead of relabeling private internals solely for this migration.

**Review by yuhezhang-ai — CHANGES_REQUESTED** [review]
Requesting changes for several correctness and maintainability issues in the new enforcement: stale mock targets make tests order-dependent, the default unit-test environment cannot collect the new contract tests, the inherited independence check is made vacuous, lazy exports erase static types, and the static __all__ evaluator can approve names that do not exist at runtime. I left inline details. The public-API relabeling and documentation scope should also be resolved deliberately. The broader import migration otherwise appears runtime-compatible.

File: docs/about/key-features.mdx
Comment (akoumpa) [reply docs/about/key-features.mdx]: Addressed in [b3490ee](https://github.com/NVIDIA-NeMo/Automodel/pull/3794/commits/b3490ee9d89179def586575e8c4a61e51e3ef8d9): [updated text](https://github.com/NVIDIA-NeMo/Automodel/blob/b3490ee9d89179def586575e8c4a61e51e3ef8d9/docs/about/key-features.mdx#L10).

Comment (akoumpa) [reply docs/about/key-features.mdx]: Addressed in [b3490ee](https://github.com/NVIDIA-NeMo/Automodel/pull/3794/commits/b3490ee9d89179def586575e8c4a61e51e3ef8d9): [updated text](https://github.com/NVIDIA-NeMo/Automodel/blob/b3490ee9d89179def586575e8c4a61e51e3ef8d9/docs/about/key-features.mdx#L12).

Comment (akoumpa) [reply docs/about/key-features.mdx]: Addressed in [b3490ee](https://github.com/NVIDIA-NeMo/Automodel/pull/3794/commits/b3490ee9d89179def586575e8c4a61e51e3ef8d9): [updated text](https://github.com/NVIDIA-NeMo/Automodel/blob/b3490ee9d89179def586575e8c4a61e51e3ef8d9/docs/about/key-features.mdx#L79).

Comment (akoumpa) [reply docs/about/key-features.mdx]: Addressed in [b3490ee](https://github.com/NVIDIA-NeMo/Automodel/pull/3794/commits/b3490ee9d89179def586575e8c4a61e51e3ef8d9): [updated text](https://github.com/NVIDIA-NeMo/Automodel/blob/b3490ee9d89179def586575e8c4a61e51e3ef8d9/docs/about/key-features.mdx#L129).

File: docs/repository-structure.mdx
Comment (akoumpa) [reply docs/repository-structure.mdx]: Addressed in [b3490ee](https://github.com/NVIDIA-NeMo/Automodel/pull/3794/commits/b3490ee9d89179def586575e8c4a61e51e3ef8d9): [updated text](https://github.com/NVIDIA-NeMo/Automodel/blob/b3490ee9d89179def586575e8c4a61e51e3ef8d9/docs/repository-structure.mdx#L3).

Comment (akoumpa) [reply docs/repository-structure.mdx]: Addressed in [b3490ee](https://github.com/NVIDIA-NeMo/Automodel/pull/3794/commits/b3490ee9d89179def586575e8c4a61e51e3ef8d9): [updated text](https://github.com/NVIDIA-NeMo/Automodel/blob/b3490ee9d89179def586575e8c4a61e51e3ef8d9/docs/repository-structure.mdx#L20).

Comment (akoumpa) [reply docs/repository-structure.mdx]: Addressed in [b3490ee](https://github.com/NVIDIA-NeMo/Automodel/pull/3794/commits/b3490ee9d89179def586575e8c4a61e51e3ef8d9): [updated text](https://github.com/NVIDIA-NeMo/Automodel/blob/b3490ee9d89179def586575e8c4a61e51e3ef8d9/docs/repository-structure.mdx#L103).

Comment (akoumpa) [reply docs/repository-structure.mdx]: Addressed in [b3490ee](https://github.com/NVIDIA-NeMo/Automodel/pull/3794/commits/b3490ee9d89179def586575e8c4a61e51e3ef8d9): [updated text](https://github.com/NVIDIA-NeMo/Automodel/blob/b3490ee9d89179def586575e8c4a61e51e3ef8d9/docs/repository-structure.mdx#L117).

File: docs/about/key-features.mdx (hunk 1 above)
Comment (akoumpa) [reply docs/about/key-features.mdx]: Addressed in [b3490ee](https://github.com/NVIDIA-NeMo/Automodel/pull/3794/commits/b3490ee9d89179def586575e8c4a61e51e3ef8d9). I kept the copyedit and scoped the claim to component packages covered by the import contract so it matches what CI currently enforces: [updated wording](https://github.com/NVIDIA-NeMo/Automodel/blob/b3490ee9d89179def586575e8c4a61e51e3ef8d9/docs/about/key-features.mdx#L134).

File: docs/repository-structure.mdx
Comment (akoumpa) [reply docs/repository-structure.mdx]: Addressed in [b3490ee](https://github.com/NVIDIA-NeMo/Automodel/pull/3794/commits/b3490ee9d89179def586575e8c4a61e51e3ef8d9). I used “three primary directories” to retain the edit without implying these are the package's only top-level directories: [updated wording](https://github.com/NVIDIA-NeMo/Automodel/blob/b3490ee9d89179def586575e8c4a61e51e3ef8d9/docs/repository-structure.mdx#L17).

File: nemo_automodel/recipes/llm/train_ft.py (hunk 2 above)
Comment (akoumpa) [reply nemo_automodel/recipes/llm/train_ft.py:464]: Addressed in 69ce965da: all five stale monkeypatch targets now patch the package-level lazy export, and the related distributed test imports use the public package API.

File: tools/component_imports.py (hunk 6 above)
Comment (akoumpa) [reply tools/component_imports.py]: Addressed in 69ce965da: __all__ is evaluated from supported top-level statements in execution order, augmented updates stay synchronized, conditional mutations get path/line errors, and nested/reset/AugAssign cases are covered.

File: tools/import_linter_contracts.py (hunk 4 above)
Comment (akoumpa) [reply tools/import_linter_contracts.py]: Addressed in 69ce965da: ComponentInterfaceContract now derives directly from Contract and checks the static interface scan without mutating the import graph. The dead independence and ignore-import machinery was removed.

File: docs/about/key-features.mdx (hunk 1 above)
Comment (akoumpa) [reply docs/about/key-features.mdx]: Already addressed by b3490ee9d: the wording now explicitly scopes the claim to component packages covered by the import contract.

File: nemo_automodel/components/distributed/__init__.py (hunk 5 above)
Comment (akoumpa) [reply nemo_automodel/components/distributed/__init__.py:232]: Addressed in 69ce965da: every lazy export in all covered component packages now has a TYPE_CHECKING declaration, and the scanner enforces exact declaration-to-_LAZY_ATTRS parity with a focused regression test.

File: tests/unit_tests/test_component_imports.py (hunk 3 above)
Comment (akoumpa) [reply tests/unit_tests/test_component_imports.py]: Addressed in 69ce965da: the default scanner test module no longer imports the lint-only contract or its dependencies. The contract itself remains exercised by lint-imports.

File: nemo_automodel/components/checkpoint/__init__.py (hunk 7 above)
Comment (akoumpa) [reply nemo_automodel/components/checkpoint/__init__.py]: Addressed in 69ce965da: every public lazy mapping now targets a public implementation name. The old private names remain only as compatibility aliases to the promoted public definitions.
